[HMA] Adding HMA UI documentation and detailed walkthrough of demo functions (#1833)
facebook/ThreatExchange: doc/hma-ui.md

# Introduction to Hasher-Matcher-Actioner UI

> Open Media Match (aka Hasher-Matcher-Actioner) includes a web app where users can configure their hash banks, connect to hash exchanges like NCMEC, Stop NCII, GIFCT, and ThreatExchange, and debug matches. These features are available as demos to show the features and functionality of HMA.  

Reviewer: ignorable/fine as followup: Consider linking the third party pages for these.

I think we should probably continue to exclusively refer to HMA as HMA. Open Media Match is the name of the library (the name of which I was partial for, but we got pushback on renaming HMA).

```suggestion
Hasher-Matcher-Actioner (aka HMA) includes a web app where users can configure their signal banks, connect to exchanges like NCMEC, StopNCII, GIFCT, ThreatExchange, and others, and debug matches. These features are available as demos to show the features and functionality of HMA.  
```

> There are 4 pages in the app:

Reviewer: ignorable: This page would benefit a lot from images, but I'd prefer not to embed them in repo itself.

The workaround we used in the past was to add images to an issue or the PR, and then deep link to them in the READMEs, but I'm not sure if that's the official way, @ThisIsMissEm  might have some ideas.

Author: This is embarrassing but I forgot how to add images. @ThisIsMissEm if you think having them in the repo is best, I can figure it out!

Reviewer 2: Yeah, so it's like add an image to `doc/images/name.png` then `![alt text](/doc/images/name.png)` iirc.

You can also just use html `img` tags: `<img src="/doc/images/name.png" alt="NLNet" />`

Reviewer: Hmm, @ThisIsMissEm - just for clarity - I think it's possible to version which images are being pointed to without adding them to repo, which increases the size.

Please see this video for this trick: https://www.youtube.com/watch?v=2ajWUp8F694

I think this allows us to version the images without *checking in* the images.

Reviewer 2: The repo is already really large, a few PNGs isn't gonna make a difference:

```
$ du -sh ThreatExchange
435M	ThreatExchange
```

And that way the images are with the repo when you check it out, and you don't need to rely on a third-party host being available.

Reviewer 2: Oh, and that "image in a issue" thing, that's been a security issue in the past, I can't find the report now, but basically it was a vulnerability to have github store files forever that weren't actually referenced anywhere — essentially upload the file, copy the URL and never use the file in anything.

## **Dashboard (`/ui/`)**

> 

Reviewer: Ignorable/Idea: Include a link to the default localhost port hosted version for each of these pages.

* View system status  
* See signal types and content types  
* Check index statistics

> There are 3 demo buttons and an option to do a factory reset:

Author: fixed in latest commit

* It immediately provides workable test data

##### Demo Purpose

* **Shows System Functionality:** Demonstrates how external exchanges work without real credentials  
* **Testing Matching**: Provides known hashes you can test matching against  
* **Index Building**: Triggers the indexing process with sample data  
* **UI Validation:** Lets you see how the exchange status, fetching, and matching systems work

##### Educational Value

* **Example Implementation:** Shows how signal exchanges are structured  
* **API Pattern**: Demonstrates the exchange API interface without complexity  
* **No External Dependencies:** Works completely offline

#### **What You See in the UI:**

1. **Before Setup:** Button shows "Setup Sample API"  
2. **After Setup:** Button shows "✅ Sample API created" and is disabled  
3. **Exchange Status:** Shows SEED\_SAMPLE exchange in the exchanges page  
4. **Index Building:** Triggers automatic background indexing of the sample data  
5. **Matching Ready:** System now has data to match against

### Technical Implementation:

The StaticSampleSignalExchangeAPI is essentially a hardcoded data source that:

> * Returns a fixed set of sample hashes (PDQ and MD5)  

Reviewer 2: As someone reading this, my question would be: Where do these sample hashes come from? How do I find the images that would trigger matches on these sample hashes?

* Provides metadata and signal data that the system can index and match against  
* Requires no external network calls or authentication

## **"Seed Banks"**

Based on the code analysis and logs, the Seed Banks button:

### Creates Test Banks

* 2 Banks: SEED\_BANK\_0 and SEED\_BANK\_1  
* Full Matching: Both set to 100% matching enabled

### Generates Massive Sample Data

* 10,000 Total Hashes (5,000 per bank)  
* Mixed Types: PDQ perceptual hashes (\~5,004) \+ MD5 video hashes (\~5,000)  
* Synthetic but Realistic: Uses ThreatExchange library's random signal generators

### System Effects

* Triggers Index Rebuild: Background process takes \~30 seconds  
* Ready for Testing: Creates searchable data immediately  
* Performance Scale: Provides substantial dataset for realistic testing

### Use Cases

* Algorithm Testing: Test similarity matching algorithms  
* UI Development: Populate interface with realistic data volumes  
* Performance Testing: Benchmark system with substantial datasets  
* Demo Workflows: Enable full end-to-end testing without real uploads

The seeded data is stored in a PostgreSQL database that runs inside the dev container:

#### **Database Structure**

> * Database: media\_match  
> * Tables: bank, bank\_content, content\_signal  
> * Host: localhost:5432 (inside the dev container)  
> * Credentials: media\_match/hunter2

Reviewer 2: ```suggestion
* Database: `media_match`  
* Tables: `bank`, `bank_content`, `content_signal`  
* Host: `localhost:5432` (inside the dev container)  
* Credentials: `media_match/hunter2`
```

#### **Data Organization**

> * 2 Banks: SEED\_BANK\_0 and SEED\_BANK\_1  

Reviewer 2: ```suggestion
* 2 Banks: `SEED_BANK_0` and `SEED_BANK_1`  
```

* \~10,000 Total Hashes: 5,004 PDQ \+ 5,000 MD5  
* Each Content Item: Has a unique ID and associated signal/hash values

### “Setup Tx API"

 The “Setup Tx API” button creates a demo integration with [ThreatExchange](https://developers.facebook.com/docs/threat-exchange/). 

#### **Integration Details**

> * **Exchange Name:** TX\_EXAMPLE\_COLLAB  

Reviewer 2: ```suggestion
* **Exchange Name:** `TX_EXAMPLE_COLLAB`
```

* **API Type:** Facebook's ThreatExchange API (FBThreatExchangeSignalExchangeAPI)  
* **Target**: Privacy group ID 1012185296055235 (demo/test group)  
* **Real Connection**: Unlike Sample API, this API connects to actual ThreatExchange infrastructure (albeit a demo one)

#### **ThreatExchange Benefits**

* **Industry Scale:** Access to millions of real threat content hashes  
* **Collaborative Defense:** Shared threat intelligence across major platforms  
* **Live Updates:** Periodic sync with updated threat databases  
* **Multiple Signal Types:** PDQ, MD5, and other content signatures

#### **Technical Details**

* **Authentication Required:** Needs valid Facebook App credentials for production (more info [here](https://developers.facebook.com/docs/threat-exchange/getting-access/))  
* **Background Fetching:** Automatic periodic synchronization  

> * **Bank Integration:** Imported data goes into dedicated bank  

Reviewer: This should be true for all exchanges that implement the interface

> * **API Limits:** Subject to ThreatExchange rate limiting

Reviewer: (Spoilers: The rate limits are high enough that no one will hit them)

Reviewer: I think removing the reference to rate limiting is probably the easiest.

#### **Important Caveats**

* **Demo Mode:** Uses test privacy group that may not have live data  
* **Credentials Needed:** Won't fetch real data without valid API credentials, which users can apply for access for [here](https://developers.facebook.com/docs/threat-exchange/getting-access/).  
* **Production Setup:** Real deployments need proper app registration with Facebook for Developers.

## **Banks (`/ui/banks`)**

> * See your test banks: `SEED_BANK_0` and `SEED_BANK_1`  

Reviewer 2: ```suggestion
* See your content banks, e.g., the `SEED_BANK_0` and `SEED_BANK_1` that are created using the demo buttons.
```

* Each contains 5,000 sample hashes  
* Try uploading images to test matching for content in hash banks

### Creating a Demo Bank

Users can create a hash bank separate from external hash exchanges like the NCMEC hash list or GIFCT hash list. Simply click “Create New Bank” and manually upload a few files to test.

Once you add content, you can test by uploading either the same or different image in the “Find Content in Banks” section. If there is a match, the UI will show which hash bank the content was found in.

### Browsing Hash Banks

The web UI only allows searching for specific files within the hash banks. If users want to explore an overview or browse content in the hash banks, they have to query the database directly.

#### **Database Table Stats (Quickest Overview)**

Run this command to see high-level statistics:

```shell
flask --app OpenMediaMatch.app table_stats
```

This shows counts of banks, contents, signals, etc.

#### **Direct Database Queries**

You can also connect to PostgreSQL directly to browse the data:

```shell
# Connect to database  

> psql -h localhost -p 5432 -U media_match -d media_match

Reviewer 2: ```suggestion
psql -h localhost -p 5432 -U media_match -d media_match
# Enter password: hunter2
```

# List all banks 
SELECT * FROM bank;
# See all content in a specific bank 
SELECT bc.id, bc.bank_id, b.name, cs.signal_type, cs.signal_val 
FROM bank_content bc 
JOIN bank b ON bc.bank_id = b.id 
JOIN content_signal cs ON bc.id = cs.content_id 
WHERE b.name = 'SEED_BANK_0';
# Count contents per bank 
SELECT b.name, COUNT(bc.id) as content_count 
FROM bank b 
LEFT JOIN bank_content bc ON b.id = bc.bank_id 
GROUP BY b.name;
```

#### **REST API Endpoints**

The API provides several endpoints to browse banks:  
**List all banks:**

```shell
curl http://localhost:5000/c/banks
```

**Get specific bank info:**

```shell
curl http://localhost:5000/c/bank/SEED_BANK_0
```

**Get specific content details:**

```shell
curl http://localhost:5000/c/bank/SEED_BANK_0/content/1
```

## **Exchanges (`/ui/exchanges`)**

The Exchanges page lets users configure external hash exchanges that are API based.

> ### Supported Hash Exchanges

Reviewer: blocking: We should link to https://github.com/facebook/ThreatExchange/tree/main/python-threatexchange/threatexchange/extensions

or 

https://github.com/facebook/ThreatExchange/blob/main/python-threatexchange/threatexchange/exchanges
```suggestion
### Default Signal Exchanges
```

**NCMEC (National Center for Missing & Exploited Children)**

* Child safety hash sharing program  
* Requires authentication credentials  
* Automated hash synchronization

**GIFCT (Global Internet Forum to Counter Terrorism)**

* Terrorist content hash database  
* Enterprise-level access required  
* Real-time threat intelligence updates

**Meta ThreatExchange**

* Comprehensive threat intelligence platform  
* API key authentication required  
* Multi-signal type support

**StopNCII**

* Non-consensual intimate image hashes  
* Privacy-focused hash sharing  
* Victim-centered approach

### Exchange Configuration

> Several of the hash exchanges supported require credentials, provided in the form of a JSON blob:

Reviewer: blocking: Link to the pytx repo on further instruction: 

The JSON represents the [CollabConfig](https://github.com/facebook/ThreatExchange/blob/main/python-threatexchange/threatexchange/exchanges/collab_config.py#L12) for the exchange. The keys and data types must match the structure data. For example, here's the exchange config for NMCEC: https://github.com/facebook/ThreatExchange/blob/main/python-threatexchange/threatexchange/exchanges/impl/ncmec_api.py#L101-L119

From that, we can see it has two fields:
1. [Required] A field named environment which is a string enum for the API endpoint
2. An optional field only_esp_ids, which accepts a container of ints

So one valid json blob would be
```
{
  'environment': 'https://hashsharing.ncmec.org/npo'
}
```

And one that filled out the optional field would be:
```
{
  'environment': 'https://hashsharing.ncmec.org/npo',
  'only_esp_ids': [1]
}
```

@ThisIsMissEm - Here's the automatic type-correct parsing and documentation reflection attempt I told you about - it could be adapted for a UI form:
https://github.com/facebook/ThreatExchange/blob/main/python-threatexchange/threatexchange/cli/config_cmd.py#L95-L101
```suggestion
Several of the exchanges supported require configuration that match their [CollabConfig](https://github.com/facebook/ThreatExchange/blob/main/python-threatexchange/threatexchange/exchanges/collab_config.py). For now, these must be provided in JSON form - consider using the [pytx CLI](https://github.com/facebook/ThreatExchange/tree/main/python-threatexchange#threatexchange-cli) to help with this. Set up a collaboration on the CLI with `threatexchange collab edit` and then print the JSON to paste in with `threatexchange config collab print`
```

Reviewer 2: @Dcallies yeah, that's the similar approach I took for the UI that I demoed in the discord, before I abandoned it because it got very complex very quick. Would rather have a json schema to be honest for these.

Reviewer 2: I also found in testing that `only_esp_ids` couldn't be missing, but had to be present and contain at least one item. But maybe I was doing something wrong.

Author: ack! I will add this in.

> 1. Configure authentication credentials (API keys, certificates)  

Reviewer: We might need to give a hint on the authentication configuration. This might also be a documentation need for us on the exchanges themselves - we might want to include a future README.md in https://github.com/facebook/ThreatExchange/tree/main/python-threatexchange/threatexchange/exchanges which includes the collaboration config explainer and which credentials are accepted, which hopefully is just copying docs from the code.

2. Select signal types to synchronize  
3. Set fetch frequency and retry policies  
4. Monitor connection health and data flow

> 

Reviewer: ignorable: You could add a mention that exchanges can be added to HMA by adding an extension: https://github.com/facebook/ThreatExchange/tree/main/python-threatexchange/threatexchange/extensions
